File: src/main/java/ru/algocode/exam2022/GameState.java
package ru.algocode.exam2022;

import org.bukkit.*;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scoreboard.DisplaySlot;
import org.bukkit.scoreboard.Objective;
import org.bukkit.scoreboard.Scoreboard;
import org.bukkit.scoreboard.ScoreboardManager;
import ru.algocode.ejudge.EjudgeSession;
import ru.algocode.ejudge.EjudgeSessionException;

import java.util.*;

import static java.lang.Math.abs;
import static java.lang.Math.log;

class GameState {
    private JavaPlugin plugin;
    private Config config;

    private SheetsAPI playersApi;

    private HashMap<String, Stats> players;
    private int playersCount;

    private Objective scoreboardObjective;

    private int secondsCount;

    ExternalXmlParser xmlParser;

    GameState(JavaPlugin plugin) {
        this.plugin = plugin;
        xmlParser = new ExternalXmlParser(plugin);
        playersCount = 1;
        secondsCount = 0;
        playersApi = new SheetsAPI(plugin.getConfig().getString("spreadsheet_id"));
        config = new Config(plugin);
        players = new HashMap<>();
        ReloadConfig();

        try {
            xmlParser.UpdateStandings(config.ExternalXmlPath, players);
        } catch (Exception e) {
            System.out.println("Failed to update standings");
        }
        reloadPlayers();
        tickPlayers();

        ScoreboardManager manager = Bukkit.getScoreboardManager();
        Scoreboard board = manager.getNewScoreboard();
        scoreboardObjective = board.registerNewObjective("score", "dummy");
        scoreboardObjective.setDisplaySlot(DisplaySlot.SIDEBAR);
        scoreboardObjective.setDisplayName(ChatColor.GOLD + "Таблица результатов");
    }

    void Tick() {
        secondsCount++;

        if (secondsCount % 60 == 0) {
            reloadPlayers();
        }

        if (secondsCount % 20 == 10) {
            try {
                xmlParser.UpdateStandings(config.ExternalXmlPath, players);
            } catch (Exception e) {
                System.out.println("Failed to update standings");
            }
        }

        tickPlayers();
        spawnItems();
    }

    void InitPlayer(Player player) {
        player.setScoreboard(scoreboardObjective.getScoreboard());
        updatePlayerInfo(player);
    }

    void Died(Player player) {
        String name = player.getName();
        if (players.containsKey(name)) {
            players.get(name).Died();
        }
        updatePlayerInfo(player);
    }

    void Killed(Player player) {
        String name = player.getName();
        if (players.containsKey(name)) {
            Stats playerStats = players.get(name);
            if (playerStats.GetTimeUntilNextKill() > 0) {
                playerStats.CruelKilledSomeone(config.TimeUntilNextKill);
                player.sendMessage("Слишком частые убийства");
            } else {
                playerStats.KilledSomeone(config.TimeUntilNextKill);
            }
        }
        updatePlayerInfo(player);
    }

    void FillChest(Inventory inventory) {
        for (ParametrizedItemStack item : config.ChestItems) {
            if (item.Generated()) {
                ItemStack itemStack = item.GenerateItem();
                if (item != null) {
                    int pos = Utils.random.nextInt(inventory.getSize());
                    inventory.setItem(pos, itemStack);
                }
            }
        }
    }

    void OpenMerchant(Player player) {
        config.MerchantMenu.Open(player);
    }

    void BuyMerchant(InventoryClickEvent event) {
        ParametrizedItemStack itemStack = config.MerchantMenu.OnClick(event);
        if (itemStack == null) {
            return;
        }

        ItemStack item = itemStack.GenerateItem();
        ItemMeta meta = item.getItemMeta();
        meta.setLore(Collections.emptyList());
        item.setItemMeta(meta);

        Player player = (Player) event.getWhoClicked();

        if (player.isOp()) {
            player.getInventory().addItem(item);
            return;
        }
        String name = player.getName();
        if (!players.containsKey(name)) {
            return;
        }

        Stats stats = players.get(name);
        if (stats.GetScore() >= itemStack.GetParam()) {
            stats.Purchase(itemStack.GetParam());
            player.getInventory().addItem(item);
            updatePlayerInfo(player);
        } else {
            Utils.sendTitle(player, ChatColor.RED + "Недостаточно денег!");
        }
    }

    void ReloadConfig() {
        int newPlayersCount = config.ReloadConfig();

        for (List<Object> row : playersApi.get("Gamestats!A" + (playersCount + 1) + ":M" + newPlayersCount)) {
            Stats player = new Stats(row);
            players.put(player.GetLogin(), player);
        }

       for (List<Object> row : playersApi.get("Problems!B" + (playersCount + 1) + ":" + (char)('B' + config.ProblemsCount) + newPlayersCount)) {
            String login = (String) row.get(0);
            if (players.containsKey(login)) {
                players.get(login).LoadProblems(row.subList(1, row.size()));
                players.get(login).RecalculateScore(config.StatsMultiplier);
            }
        }

        for (List<Object> ejudgeAuth : config.GetSheetsApi().get("Users!A" + (playersCount + 1) + ":E" + newPlayersCount)) {
            String login = (String) ejudgeAuth.get(1);
            if (players.containsKey(login)) {
                players.get(login).AddEjudgeAuth(ejudgeAuth);
            }
        }

        playersCount = newPlayersCount;
    }

    void SubmitProblem(Player player, String label, String code) {
        String name = player.getName();
        if (players.containsKey(name)) {
            Stats playerStats = players.get(name);

            Thread t = new Thread(new Runnable() {
                String login = playerStats.GetEjudgeLogin();
                String pasword = playerStats.GetEjudgePassword();
                String problem = label;
                String source = code;
                int contest = config.EjudgeContestId;
                public void run() {
                    EjudgeSession session = new EjudgeSession(login, pasword, contest);
                    session.authenticate();
                    session.submit(problem, source);
                }
            });
            t.start();
        }
    }

    void GetStatus(Player player) {
        String name = player.getName();
        if (players.containsKey(name)) {
            Stats playerStats = players.get(name);
            player.sendMessage(playerStats.GetStatus());
        }
    }

    private void spawnItems() {
        for (ParametrizedItemStack item : config.SpawnItems) {
            if (item.Generated()) {
                ItemStack itemStack = item.GenerateItem();
                if (item != null) {
                    World world = plugin.getServer().getWorld("world");
                    int x = config.SpawnerX + Utils.random.nextInt(config.SpawnerRadius * 2 + 1) - config.SpawnerRadius - 1;
                    int z = config.SpawnerZ + Utils.random.nextInt(config.SpawnerRadius * 2 + 1) - config.SpawnerRadius - 1;
                    Location loc = new Location(world, x, config.SpawnerY, z);
                    loc.getWorld().dropItemNaturally(loc, itemStack);
                }
            }
        }
    }

    private void tickPlayers() {
        for (Player player : plugin.getServer().getOnlinePlayers()) {
            String name = player.getName();
            if (players.containsKey(name)) {
                Stats playerStats = players.get(name);
                playerStats.Tick();
                if (secondsCount % 60 == 0) {
                    playerStats.IncInGame();
                }
                Location location = player.getLocation();
                if (
                        abs(location.getX() - config.SpawnerX) <= config.SpawnerRadius * 2 &&
                        abs(location.getY() - config.SpawnerY) <= config.SpawnerRadius * 2 &&
                        abs(location.getZ() - config.SpawnerZ) <= config.SpawnerRadius * 2
                ) {
                    if (!playerStats.GetIsInForbiddenZone()) {
                        player.sendMessage("Вы в запретной зоне, покинте её в течении " +
                                (config.MaxTimeInForbiddenZone - playerStats.GetTimeInForbiddenZone()) / config.TimeIncreaseForForbiddenZone +
                                "секунд"
                        );
                    }
                    playerStats.InForbiddenZone(config.TimeIncreaseForForbiddenZone);
                    if (playerStats.GetTimeInForbiddenZone() >= config.MaxTimeInForbiddenZone) {
                        player.sendMessage(ChatColor.RED + "Слишком долго в запретной зоне, штраф " + config.PenaltyForForbiddenZone);
                        playerStats.TooLongInForbiddenZone(config.PenaltyForForbiddenZone);
                    }
                } else {
                    playerStats.InAllowedZone();
                }
                updatePlayerInfo(player);
            }
        }
    }

    private void reloadPlayers() {
        List<List<Object>> gamestats = new ArrayList<>();
        List<List<Object>> problems = new ArrayList<>();

        int idx = 0;
        for (List<Object> row : playersApi.get("Gamestats!B2:J" + playersCount)) {
            gamestats.add(new ArrayList<>());
            problems.add(new ArrayList<>());
            String login = (String) row.get(0);
            int bonus = Integer.parseInt((String) row.get(8));

            if (players.containsKey(login)) {
                players.get(login).SetBonus(bonus);
                players.get(login).RecalculateScore(config.StatsMultiplier);
                players.get(login).ExportStats(gamestats.get(idx), problems.get(idx));
            }
            idx++;
        }

        playersApi.update("Gamestats!C2:M" + playersCount, gamestats);
        playersApi.update("Problems!C2:" + (char)('C' + config.ProblemsCount) + playersCount, problems);
    }

    private void updatePlayerInfo(Player player) {
        String name = player.getName();
        String displayName;
        if (player.isOp()) {
            displayName = ChatColor.RED + "Преподаватель " + ChatColor.GOLD + name + ChatColor.RESET;
        } else if (players.containsKey(name)) {
            Stats playerStats = players.get(name);
            playerStats.RecalculateScore(config.StatsMultiplier);
            int score = playerStats.GetScore();
            displayName = "(" + score + ") " + ChatColor.GREEN + playerStats.GetName() + ChatColor.RESET;
            scoreboardObjective.getScore(playerStats.GetName()).setScore(score);
        } else {
            displayName = name;
        }

        player.setDisplayName(displayName);
        player.setPlayerListName(displayName);
    }

}

File: src/main/java/ru/algocode/exam2022/GameState.kt
package ru.algocode.exam2022

import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Location
import org.bukkit.boss.BarColor
import org.bukkit.boss.BarStyle
import org.bukkit.boss.BossBar
import org.bukkit.entity.Player
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.Inventory
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.scoreboard.DisplaySlot
import org.bukkit.scoreboard.Objective
import ru.algocode.ejudge.EjudgeSession
import ru.algocode.exam2022.utils.ExternalXmlParser
import ru.algocode.exam2022.utils.SheetsAPI
import ru.algocode.exam2022.utils.Stats
import ru.algocode.exam2022.utils.Utils
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.abs
import kotlin.math.max

class GameState internal constructor(private val plugin: JavaPlugin) {
    val config: Config
    private val playersApi: SheetsAPI
    private val players: ConcurrentHashMap<String, Stats>
    private val bossBars: ConcurrentHashMap<String, BossBar> = ConcurrentHashMap()
    private var playersCount: Int
    private val scoreboardObjective: Objective
    private var secondsCount: Int
    var xmlParser: ExternalXmlParser = ExternalXmlParser(plugin)

    init {
        playersCount = 1
        secondsCount = 0
        playersApi = SheetsAPI(plugin.config.getString("spreadsheet_id")!!, plugin)
        config = Config(plugin)
        players = ConcurrentHashMap()
        reloadConfig()
        try {
            xmlParser.UpdateStandings(config.externalXmlPath, players)
        } catch (e: Exception) {
            println("Failed to update standings")
        }
        reloadPlayers()
        tickPlayers()
        val manager = Bukkit.getScoreboardManager()
        val board = manager!!.newScoreboard
        scoreboardObjective =
            board.registerNewObjective(
                "score",
                "dummy",
                ChatColor.GOLD.toString() +
                        "Таблица результатов",
            )
        scoreboardObjective.displaySlot = DisplaySlot.SIDEBAR
    }

    fun tick() {
        secondsCount++
        if (secondsCount % 60 == 0) {
            reloadPlayers()
        }
        if (secondsCount % 20 == 10) {
            try {
                xmlParser.UpdateStandings(config.externalXmlPath, players)
            } catch (e: Exception) {
                println("Failed to update standings")
            }
        }
        tickPlayers()
        spawnItems()
    }

    fun initPlayer(player: Player) {
        player.scoreboard = scoreboardObjective.scoreboard!!
        updatePlayerInfo(player)
    }

    fun died(player: Player) {
        val name = player.name
        if (players.containsKey(name)) {
            players[name]!!.Died()
        }
        updatePlayerInfo(player)
    }

    fun killed(player: Player) {
        val name = player.name
        if (players.containsKey(name)) {
            val playerStats = players[name]
            if (playerStats!!.GetTimeUntilNextKill() > 0) {
                playerStats.CruelKilledSomeone(config.timeUntilNextKill)
                player.sendMessage("Слишком частые убийства")
            } else {
                playerStats.KilledSomeone(config.timeUntilNextKill)
            }
        }
        updatePlayerInfo(player)
    }

    fun fillChest(inventory: Inventory) {
        for (item in config.chestItems!!) {
            if (item.Generated()) {
                val itemStack = item.GenerateItem()
                val pos = Utils.random.nextInt(inventory.size)
                inventory.setItem(pos, itemStack)
            }
        }
    }

    fun openMerchant(player: Player?) {
        config.merchantMenu!!.Open(player)
    }

    fun buyMerchant(event: InventoryClickEvent) {
        val itemStack = config.merchantMenu!!.OnClick(event) ?: return
        val item = itemStack.GenerateItem()
        val meta = item.itemMeta
        meta!!.lore = emptyList()
        item.itemMeta = meta
        val player = event.whoClicked as Player
        if (player.isOp) {
            player.inventory.addItem(item)
            return
        }
        val name = player.name
        if (!players.containsKey(name)) {
            return
        }
        val stats = players[name]
        if (stats!!.GetScore() >= itemStack.GetParam()) {
            stats.Purchase(itemStack.GetParam())
            player.inventory.addItem(item)
            updatePlayerInfo(player)
        } else {
            Utils.sendTitle(player, ChatColor.RED.toString() + "Недостаточно денег!")
        }
    }

    fun reloadConfig() {
        val newPlayersCount = config.ReloadConfig()
        for (row in playersApi["Gamestats!A" + (playersCount + 1) + ":M" + newPlayersCount]!!) {
            val player = Stats(row)
            players[player.GetLogin()] = player
        }
        for (row in playersApi["Problems!B" + (playersCount + 1) + ":" + ('B'.code + config.problemsCount).toChar() + newPlayersCount]!!) {
            val login = row[0] as String
            if (players.containsKey(login)) {
                players[login]!!.LoadProblems(row.subList(1, row.size))
                players[login]!!.RecalculateScore(config.statsMultiplier)
            }
        }
        for (ejudgeAuth in config.GetSheetsApi()["Users!A" + (playersCount + 1) + ":E" + newPlayersCount]!!) {
            val login = ejudgeAuth[1] as String
            if (players.containsKey(login)) {
                players[login]!!.AddEjudgeAuth(ejudgeAuth)
            }
        }
        playersCount = newPlayersCount
    }

    fun submitProblem(player: Player, label: String, code: String) {
        val name = player.name
        if (players.containsKey(name)) {
            val playerStats = players[name]
            val t = Thread(object : Runnable {
                var login = playerStats!!.GetEjudgeLogin()
                var pasword = playerStats!!.GetEjudgePassword()
                var problem = label
                var source = code
                var contest = config.ejudgeContestId
                override fun run() {
                    val session = EjudgeSession(login, pasword, contest)
                    session.authenticate()
                    session.submit(problem, source)
                }
            })
            t.start()
        }
    }

    fun GetStatus(player: Player) {
        val name = player.name
        if (players.containsKey(name)) {
            val playerStats = players[name]
            player.sendMessage(playerStats!!.GetStatus())
        }
    }

    private fun spawnItems() {
        for (item in config.spawnItems!!) {
            if (item.Generated()) {
                val itemStack = item.GenerateItem()
                val world = plugin.server.getWorld("world")
                val x =
                    config.spawnerX + Utils.random.nextInt(config.spawnerRadius * 2 + 1) - config.spawnerRadius - 1
                val z =
                    config.spawnerZ + Utils.random.nextInt(config.spawnerRadius * 2 + 1) - config.spawnerRadius - 1
                val loc = Location(world, x.toDouble(), config.spawnerY.toDouble(), z.toDouble())
                loc.world!!.dropItemNaturally(loc, itemStack)
            }
        }
    }

    private fun tickPlayers() {
        for (player in plugin.server.onlinePlayers) {
            val name = player.name
            if (players.containsKey(name)) {
                val playerStats = players[name]
                playerStats!!.Tick()
                if (secondsCount % 60 == 0) {
                    playerStats.IncInGame()
                }
                val location = player.location
                if (
                    abs(location.x - config.spawnerX) <= config.spawnerRadius &&
                    abs(location.z - config.spawnerZ) <= config.spawnerRadius
                ) {
                    if (!playerStats.GetIsInForbiddenZone()) {
                        bossBars[name]?.removeAll()
                        bossBars.remove(name)
                        bossBars[name] =
                            plugin.server.createBossBar("Запретная зона", BarColor.RED, BarStyle.SOLID).apply {
                                addPlayer(player)
                                progress = max(
                                    1.0 - playerStats.GetTimeInForbiddenZone() / config.maxTimeInForbiddenZone.toDouble(),
                                    0.0
                                )
                                isVisible = true
                            }
                    }
                    playerStats.InForbiddenZone(config.timeIncreaseForForbiddenZone)
                    if (playerStats.GetTimeInForbiddenZone() >= config.maxTimeInForbiddenZone) {
                        val scoreBefore = playerStats.GetScore()
                        playerStats.TooLongInForbiddenZone(config.penaltyForForbiddenZone)
                        bossBars[name]!!.run {
                            color = BarColor.YELLOW
                            setTitle(
                                "Запретная зона: штраф " +
                                        "${config.penaltyForForbiddenZone}, " +
                                        "$scoreBefore -> ${scoreBefore - config.penaltyForForbiddenZone}"
                            )
                        }
                    } else
                        bossBars[name]!!.run {
                            progress = max(
                                1.0 - playerStats.GetTimeInForbiddenZone() / config.maxTimeInForbiddenZone.toDouble(),
                                0.0
                            )
                            setTitle(
                                "Запретная зона: осталоcь " +
                                        "${
                                            (config.maxTimeInForbiddenZone - playerStats.GetTimeInForbiddenZone())
                                                    / config.timeIncreaseForForbiddenZone
                                        } сек. "
                            )
                        }
                } else {
                    playerStats.InAllowedZone()
                    val regenSucc = "Время в запретной зоне восстановлено"
                    if (playerStats.GetTimeInForbiddenZone() == 0 && bossBars[name]?.title == regenSucc) {
                        bossBars[name]?.removeAll()
                        bossBars.remove(name)
                    } else if (bossBars[name] != null) {
                        bossBars[name]!!.run {
                            if (playerStats.GetTimeInForbiddenZone() > 0) {
                                color = BarColor.GREEN
                                progress = max(
                                    1.0 -
                                            playerStats.GetTimeInForbiddenZone()
                                            / config.maxTimeInForbiddenZone.toDouble(),
                                    0.0
                                )
                                val left = "Вы покинули запретную зону"
                                val regen = "Восстановление:"

                                setTitle(
                                    if (title == left) {
                                        regen
                                    } else {
                                        if (title == regen || title.isEmpty()) {
                                            isVisible = false
                                            ""
                                        } else {
                                            left
                                        }
                                    }
                                )
                            } else {
                                isVisible = true
                                setTitle(regenSucc)
                            }
                        }
                    }
                }
                updatePlayerInfo(player)
            }
        }
    }

    private fun reloadPlayers() {
        val gamestats: MutableList<List<Any>> = ArrayList()
        val problems: MutableList<List<Any>> = ArrayList()
        var idx = 0
        for (row in playersApi["Gamestats!B2:J$playersCount"]!!) {
            gamestats.add(ArrayList())
            problems.add(ArrayList())
            val login = row[0] as String
            val bonus: Int = (row[8] as String).toInt()
            if (players.containsKey(login)) {
                players[login]!!.SetBonus(bonus)
                players[login]!!.RecalculateScore(config.statsMultiplier)
                players[login]!!.ExportStats(gamestats[idx], problems[idx])
            }
            idx++
        }
        playersApi.update("Gamestats!C2:M$playersCount", gamestats)
        playersApi.update("Problems!C2:" + ('C'.code + config.problemsCount).toChar() + playersCount, problems)
    }

    private fun updatePlayerInfo(player: Player) {
        val name = player.name
        val displayName: String
        if (player.isOp) {
            displayName = ChatColor.RED.toString() + "Преподаватель " + ChatColor.GOLD + name + ChatColor.RESET
        } else if (players.containsKey(name)) {
            val playerStats = players[name]
            playerStats!!.RecalculateScore(config.statsMultiplier)
            val score = playerStats.GetScore()
            displayName = "(" + score + ") " + ChatColor.GREEN + playerStats.GetName() + ChatColor.RESET
            scoreboardObjective.getScore(playerStats.GetName()).score = score
        } else {
            displayName = name
        }
        player.setDisplayName(displayName)
        player.setPlayerListName(displayName)
    }
}